tendermint/src/block/header.rs

```diff
@@ -70,6 +71,106 @@ pub struct Header {
+        let mut last_block_id_enc = vec![];
+        amino_types::BlockId::from(&self.last_block_id)
+            .encode(&mut last_block_id_enc)
+            .unwrap();
+        let mut last_commit_hash_enc = vec![];
+        if let Some(last_commit_hash_bytes) = self.last_commit_hash.as_bytes() {
+            last_commit_hash_enc = encode_bytes(last_commit_hash_bytes);
+        }
+        let mut data_hash_enc = vec![];
+        if let Some(data_hash_bytes) = self.data_hash.as_bytes() {
+            data_hash_enc = encode_bytes(data_hash_bytes);
+        }
+        let mut validator_hash_enc = vec![];
+        if let Some(validator_hash_bytes) = self.validators_hash.as_bytes() {
+            validator_hash_enc = encode_bytes(validator_hash_bytes);
+        }
+        let mut next_validator_hash_enc = vec![];
+        if let Some(next_validator_hash_bytes) = self.next_validators_hash.as_bytes() {
+            next_validator_hash_enc = encode_bytes(next_validator_hash_bytes);
+        }
+        let mut consensus_hash_enc = vec![];
+        if let Some(consensus_hash_bytes) = self.consensus_hash.as_bytes() {
+            consensus_hash_enc = encode_bytes(consensus_hash_bytes);
+        }
+        let mut app_hash_enc = vec![];
+        if let Some(app_hash_bytes) = self.app_hash.as_bytes() {
+            app_hash_enc = encode_bytes(app_hash_bytes);
+        }
+        let mut last_result_hash_enc = vec![];
+        if let Some(last_result_hash_bytes) = self.last_results_hash.as_bytes() {
+            last_result_hash_enc = encode_bytes(last_result_hash_bytes);
+        }
+        let mut evidence_hash_enc = vec![];
+        if let Some(evidence_hash_bytes) = self.evidence_hash.as_bytes() {
+            evidence_hash_enc = encode_bytes(evidence_hash_bytes);
+        }
+        let proposer_address_bytes = self.proposer_address.as_bytes();
+        let proposer_address_enc = encode_bytes(&proposer_address_bytes);
+
+        let mut byteslices: Vec<&[u8]> = vec![];
+        byteslices.push(version_enc.as_slice());
+        byteslices.push(chain_id_enc.as_slice());
+        byteslices.push(height_enc.as_slice());
+        byteslices.push(time_enc.as_slice());
+        byteslices.push(num_tx_enc.as_slice());
+        byteslices.push(total_tx_enc.as_slice());
+        byteslices.push(last_block_id_enc.as_slice());
+        byteslices.push(last_commit_hash_enc.as_slice());
+        byteslices.push(data_hash_enc.as_slice());
+        byteslices.push(validator_hash_enc.as_slice());
+        byteslices.push(next_validator_hash_enc.as_slice());
+        byteslices.push(consensus_hash_enc.as_slice());
+        byteslices.push(app_hash_enc.as_slice());
+        byteslices.push(last_result_hash_enc.as_slice());
+        byteslices.push(evidence_hash_enc.as_slice());
+        byteslices.push(proposer_address_enc.as_slice());
+
+        Hash::Sha256(simple_hash_from_byte_slices(byteslices.as_slice()))
```

Comment:
This looks a bit too exhaustive but aims to do exactly the same as: https://github.com/tendermint/tendermint/blob/134fe2896275bb926b49743c1e25493f6b24cc31/types/block.go#L393

This could probably be simplified substantially. Maybe, `simple_hash_from_byte_slices` had a similar logic as a Hasher in golang (`Write` the slices and then `Sum` to create the tree once and return the hash)?

Reply:
Definitely seems excessive.

Good idea re a write based API. Though I don't think you usually see that for Merkle trees? At least the Merkle tree version needs to track the boundaries of each write, where as a normal hasher just concatenates everything together.

Reply:
Note to myself: add back the test that checks the hash here:
https://github.com/interchainio/tendermint-rs/blob/262e252a04cb7d7b660a829f3240e7196ad955b9/tendermint/tests/rpc.rs#L167-L168

Reply:
done in https://github.com/interchainio/tendermint-rs/pull/36/commits/e686c457e2913c11737e65305a146a216dc66729